templates/instanceha/bin/instanceha.py

```diff
@@ -498,6 +498,41 @@ def _redfish_reset(url, user, passwd, timeout, action):
     r = requests.post(url, data=json.dumps(payload), headers=headers, auth=(user, passwd), verify=False, timeout=timeout)
     return r
 
+def _bmh_fence(token, namespace, host, action):
+
+    url = "https://kubernetes.default.svc/apis/metal3.io/v1alpha1/namespaces/%s/baremetalhosts/%s?fieldManager=kubectl-patch" % (namespace, host)
+    headers={'Authorization': 'Bearer '+token, 'Content-Type': 'application/merge-patch+json'}
+    cacert = '/var/run/secrets/kubernetes.io/serviceaccount/ca.crt'
+
+    if action == 'off':
+        ann={"metadata":{"annotations":{"reboot.metal3.io/iha":"{\"mode\": \"hard\"}"}}}
+        r = requests.patch(url, headers=headers, verify=cacert, data=json.dumps(ann))
+        if r.status_code == 200:
+            #check if server is off - wait up to 30s - TODO this may need tuning
+
+            headers={'Authorization': 'Bearer '+token}
+            url = "https://kubernetes.default.svc/apis/metal3.io/v1alpha1/namespaces/%s/baremetalhosts/%s" % (namespace, host)
+
+            break_loop = False
+            i=0
+            while not break_loop:
+              time.sleep(3)
+              s = requests.get(url, headers=headers, verify=cacert)
+              poweredon = json.loads(s.text)['status']['poweredOn']
+              i += 1
+              if i > 9 or not poweredon:
+                break_loop = True
+            if not poweredon:
+                return True
+            else:
+                return False
+
+        else:
+            return False
+    else:
+        ann={"metadata":{"annotations":{"reboot.metal3.io/iha":None}}}
+        r = requests.patch(url, headers=headers, verify=cacert, data=json.dumps(ann))
+        return r
```

Comment:
for the other agents we parse the response afterwards, so I kinda just let it as it is for the others, this can be improved for sure (try/except the request etc). we don't care about the poweron as much as the poweroff in general but I agree this could be improved.

Reply:
Just seems slightly evil to change the return type based on the input :-) It's relatively self contained anyway so not a big issue if/when refactoring